pkg/webhook/server/server.go

```diff
@@ -0,0 +1,295 @@
+/*
+Copyright 2019 The Jetstack cert-manager contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package server
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/go-logr/logr"
+	admissionv1beta1 "k8s.io/api/admission/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/runtime/serializer/json"
+
+	"github.com/jetstack/cert-manager/pkg/webhook/handlers"
+)
+
+var (
+	defaultScheme = runtime.NewScheme()
+)
+
+func init() {
+	admissionv1beta1.AddToScheme(defaultScheme)
+
+	// we need to add the options to empty v1
+	// TODO fix the server code to avoid this
+	metav1.AddToGroupVersion(defaultScheme, schema.GroupVersion{Version: "v1"})
+
+	// TODO: keep the generic API server from wanting this
+	unversioned := schema.GroupVersion{Group: "", Version: "v1"}
+	defaultScheme.AddUnversionedTypes(unversioned,
+		&metav1.Status{},
+		&metav1.APIVersions{},
+		&metav1.APIGroupList{},
+		&metav1.APIGroup{},
+		&metav1.APIResourceList{},
+	)
+}
+
+type Server struct {
+	// ListenAddr is the address the HTTP server should listen on
+	// This must be specified.
+	ListenAddr string
+
+	// HealthzAddr is the address the healthz HTTP server should listen on
+	// If not specified, the healthz endpoint will not be exposed.
+	HealthzAddr string
+
+	// Scheme is used to decode/encode request/response payloads.
+	// If not specified, a default scheme that registers the AdmissionReview
+	// resource type will be used.
+	Scheme *runtime.Scheme
+
+	// If specified, the server will listen with TLS using certificates
+	// provided by this CertificateSource.
+	CertificateSource CertificateSource
+
+	ValidationWebhook handlers.ValidatingAdmissionHook
+	MutationWebhook   handlers.MutatingAdmissionHook
+
+	// Log is an optional logger to write informational and error messages to.
+	// If not specified, no messages will be logged.
+	Log logr.Logger
```

Comment:
Why are we doing this maylog behavior?

Reply:
to handle the case where a logger isn't provided - we don't have a standard `New` function so we can't put a null logger here easily. I *could* just set this field to a NullLogger at the start of `Run` if you think that's cleaner.. It is the only exported function.

Reply:
updated to set `Log` when Run is called